File: web/src/lib/socActionPack.ts
// Client-side port of src/CVEzD3FEND/actions/soc_action_pack.py — every field
// is a deterministic, read-only projection over bundle nodes/edges/indexes,
// so the SOC Action Pack page works fully offline (static-first principle).
import { getNode } from "./bundle";
import type { BundleNode, KnowledgeBundle, SocActionPack } from "./types";

const PRIORITY_BY_COVERAGE: Record<string, SocActionPack["priority"]> = {
  gap: "High",
  partial: "Medium",
  covered: "Low",
  unknown: "Medium",
};

const CONFIDENCE_BY_COVERAGE: Record<string, number> = {
  gap: 0.3,
  partial: 0.6,
  covered: 1.0,
  unknown: 0.3,
};

function safeIdFragment(nodeId: string): string {
  return nodeId.replace(/[^A-Za-z0-9_-]/g, "_");
}

export function buildSocActionPack(bundle: KnowledgeBundle, attackId: string): SocActionPack {
  const attackNode = getNode(bundle, attackId);
  if (!attackNode || attackNode.type !== "attack") {
    throw new Error(`Unknown ATT&CK technique: ${attackId}`);
  }

  const idx = bundle.indexes;
  const defendIds = idx.attack_to_defend[attackId] ?? [];
  const detections = idx.attack_to_detections[attackId] ?? [];
  const coverageStatus = idx.coverage_by_technique[attackId] ?? "unknown";
  const gapIds = idx.gaps_by_technique[attackId] ?? [];

  const nodesById = new Map<string, BundleNode>(bundle.nodes.map((n) => [n.id, n]));

  const controls: string[] = [];
  const mitigations: string[] = [];
  for (const d of defendIds) {
    for (const c of idx.defend_to_controls[d] ?? []) {
      if (!controls.includes(c)) controls.push(c);
    }
    const mitId = `MIT-${d}`;
    if (nodesById.has(mitId) && !mitigations.includes(mitId)) mitigations.push(mitId);
  }

  const socActions: string[] = [];
  for (const e of bundle.edges) {
    if (e.type === "soc_action_operationalizes_defend" && defendIds.includes(e.target)) {
      if (!socActions.includes(e.source)) socActions.push(e.source);
    }
  }

  const evidence: string[] = [];
  const dataSources: string[] = [];
  const logSources: string[] = [];
  const detectionSet = new Set(detections);
  for (const e of bundle.edges) {
    if (e.type === "evidence_supports_detection" && detectionSet.has(e.target)) {
      if (!evidence.includes(e.source)) evidence.push(e.source);
    } else if (e.type === "data_source_enables_detection" && detectionSet.has(e.target)) {
      const srcNode = nodesById.get(e.source);
      if (srcNode?.type === "log_source" && !logSources.includes(e.source)) {
        logSources.push(e.source);
      } else if (srcNode?.type === "data_source" && !dataSources.includes(e.source)) {
        dataSources.push(e.source);
      }
    }
  }

  const huntId = `HUNT-${safeIdFragment(attackId)}`;
  const huntingHypotheses = nodesById.has(huntId) ? [huntId] : [];

  let attackPath: string[] = [attackId];
  for (const r of bundle.routes) {
    if (r.start_node.startsWith("CVE-") && r.nodes.includes(attackId)) {
      attackPath = r.nodes;
      break;
    }
  }

  const defensivePath = [...defendIds, ...controls];

  const sourceRefs: string[] = [];
  for (const nid of [attackId, ...defendIds, ...controls, ...detections]) {
    const node = nodesById.get(nid);
    if (!node) continue;
    for (const ref of node.source_refs) {
      if (!sourceRefs.includes(ref)) sourceRefs.push(ref);
    }
  }

  const executiveSummary =
    `${attackId} (${attackNode.name}) is currently '${coverageStatus}'. ` +
    `${defendIds.length} D3FEND technique(s), ${detections.length} detection opportunity(ies), ` +
    `and ${gapIds.length} open gap(s) are tracked for this technique.`;

  const technicalSummary =
    `D3FEND mappings: ${defendIds.join(", ") || "none"}. ` +
    `Controls: ${controls.join(", ") || "none"}. ` +
    `Mitigations: ${mitigations.join(", ") || "none"}. ` +
    `Detections: ${detections.join(", ") || "none"}.`;

  return {
    id: `PACK-${safeIdFragment(attackId)}`,
    title: `SOC Action Pack: ${attackNode.name} (${attackId})`,
    executive_summary: executiveSummary,
    technical_summary: technicalSummary,
    attack_path: attackPath,
    defensive_path: defensivePath,
    recommended_actions: [...controls, ...socActions],
    hunting_hypotheses: huntingHypotheses,
    detection_opportunities: detections,
    required_logs: [...dataSources, ...logSources],
    required_evidence: evidence,
    mitigations,
    gaps: gapIds,
    priority: PRIORITY_BY_COVERAGE[coverageStatus] ?? "Medium",
    confidence: CONFIDENCE_BY_COVERAGE[coverageStatus] ?? 0.3,
    source_refs: sourceRefs,
  };
}

File: src/CVEzD3FEND/actions/soc_action_pack.py
"""Deterministic SOC Action Pack generation, one per ATT&CK technique.

Every field is a read-only projection over `bundle.nodes`/`edges`/`indexes` —
no AI, no invented ids (per AI_ASSISTANCE_CONTRACT, "the graph validates, AI
proposes, human promotes"; this generator is part of "the graph validates").
"""

from __future__ import annotations

from CVEzD3FEND.models.bundle import Bundle
from CVEzD3FEND.models.graph import EdgeType, Node, NodeType
from CVEzD3FEND.models.soc import Priority, SocActionPack
from CVEzD3FEND.util import safe_id_fragment

_PRIORITY_BY_COVERAGE: dict[str, Priority] = {
    "gap": "High",
    "partial": "Medium",
    "covered": "Low",
    "unknown": "Medium",
}
_CONFIDENCE_BY_COVERAGE = {"gap": 0.30, "partial": 0.60, "covered": 1.0, "unknown": 0.30}


def build_soc_action_pack(bundle: Bundle, attack_id: str) -> SocActionPack:
    nodes_by_id: dict[str, Node] = {n.id: n for n in bundle.nodes}
    attack_node = nodes_by_id.get(attack_id)
    if attack_node is None or attack_node.type != NodeType.ATTACK:
        raise ValueError(f"Unknown ATT&CK technique: {attack_id}")

    indexes = bundle.indexes
    defend_ids: list[str] = indexes.get("attack_to_defend", {}).get(attack_id, [])
    detections: list[str] = indexes.get("attack_to_detections", {}).get(attack_id, [])
    coverage_status: str = indexes.get("coverage_by_technique", {}).get(attack_id, "unknown")
    gap_ids: list[str] = indexes.get("gaps_by_technique", {}).get(attack_id, [])

    controls: list[str] = []
    mitigations: list[str] = []
    for d in defend_ids:
        for c in indexes.get("defend_to_controls", {}).get(d, []):
            if c not in controls:
                controls.append(c)
        mit_id = f"MIT-{d}"
        if mit_id in nodes_by_id and mit_id not in mitigations:
            mitigations.append(mit_id)

    soc_actions: list[str] = []
    for e in bundle.edges:
        if e.type == EdgeType.SOC_ACTION_OPERATIONALIZES_DEFEND and e.target in defend_ids:
            if e.source not in soc_actions:
                soc_actions.append(e.source)

    evidence: list[str] = []
    data_sources: list[str] = []
    log_sources: list[str] = []
    detection_set = set(detections)
    for e in bundle.edges:
        if e.type == EdgeType.EVIDENCE_SUPPORTS_DETECTION and e.target in detection_set:
            if e.source not in evidence:
                evidence.append(e.source)
        elif e.type == EdgeType.DATA_SOURCE_ENABLES_DETECTION and e.target in detection_set:
            src_node = nodes_by_id.get(e.source)
            if src_node and src_node.type == NodeType.LOG_SOURCE and e.source not in log_sources:
                log_sources.append(e.source)
            elif src_node and src_node.type == NodeType.DATA_SOURCE and e.source not in data_sources:
                data_sources.append(e.source)

    hunt_id = f"HUNT-{safe_id_fragment(attack_id)}"
    hunting_hypotheses = [hunt_id] if hunt_id in nodes_by_id else []

    attack_path: list[str] = [attack_id]
    for r in bundle.routes:
        if r.start_node.startswith("CVE-") and attack_id in r.nodes:
            attack_path = r.nodes
            break

    defensive_path = [*defend_ids, *controls]

    source_refs: list[str] = []
    for nid in [attack_id, *defend_ids, *controls, *detections]:
        node = nodes_by_id.get(nid)
        if not node:
            continue
        for ref in node.source_refs:
            if ref not in source_refs:
                source_refs.append(ref)

    executive_summary = (
        f"{attack_id} ({attack_node.name}) is currently '{coverage_status}'. "
        f"{len(defend_ids)} D3FEND technique(s), {len(detections)} detection opportunity(ies), "
        f"and {len(gap_ids)} open gap(s) are tracked for this technique."
    )
    technical_summary = (
        f"D3FEND mappings: {', '.join(defend_ids) or 'none'}. "
        f"Controls: {', '.join(controls) or 'none'}. "
        f"Mitigations: {', '.join(mitigations) or 'none'}. "
        f"Detections: {', '.join(detections) or 'none'}."
    )

    return SocActionPack(
        id=f"PACK-{safe_id_fragment(attack_id)}",
        title=f"SOC Action Pack: {attack_node.name} ({attack_id})",
        executive_summary=executive_summary,
        technical_summary=technical_summary,
        attack_path=attack_path,
        defensive_path=defensive_path,
        recommended_actions=[*controls, *soc_actions],
        hunting_hypotheses=hunting_hypotheses,
        detection_opportunities=detections,
        required_logs=[*data_sources, *log_sources],
        required_evidence=evidence,
        mitigations=mitigations,
        gaps=gap_ids,
        priority=_PRIORITY_BY_COVERAGE.get(coverage_status, "Medium"),
        confidence=_CONFIDENCE_BY_COVERAGE.get(coverage_status, 0.30),
        source_refs=source_refs,
    )


def build_all_soc_action_packs(bundle: Bundle) -> list[SocActionPack]:
    return [
        build_soc_action_pack(bundle, n.id)
        for n in sorted(bundle.nodes, key=lambda x: x.id)
        if n.type == NodeType.ATTACK
    ]
